File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/SleepSessionRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.SleepSessionRecord
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.SleepSessionRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.SleepStageDto
import com.phamtunglam.health_connector_hc_android.pigeon.SleepStageTypeDto
import java.time.Instant
import java.time.ZoneOffset

internal fun SleepStageTypeDto.toHealthConnect(): Int = when (this) {
    SleepStageTypeDto.UNKNOWN -> SleepSessionRecord.STAGE_TYPE_UNKNOWN
    SleepStageTypeDto.AWAKE -> SleepSessionRecord.STAGE_TYPE_AWAKE
    SleepStageTypeDto.SLEEPING -> SleepSessionRecord.STAGE_TYPE_SLEEPING
    SleepStageTypeDto.OUT_OF_BED -> SleepSessionRecord.STAGE_TYPE_OUT_OF_BED
    SleepStageTypeDto.LIGHT -> SleepSessionRecord.STAGE_TYPE_LIGHT
    SleepStageTypeDto.DEEP -> SleepSessionRecord.STAGE_TYPE_DEEP
    SleepStageTypeDto.REM -> SleepSessionRecord.STAGE_TYPE_REM
    SleepStageTypeDto.IN_BED -> SleepSessionRecord.STAGE_TYPE_AWAKE_IN_BED
}

internal fun Int.toSleepStageTypeDto(): SleepStageTypeDto = when (this) {
    SleepSessionRecord.STAGE_TYPE_UNKNOWN -> SleepStageTypeDto.UNKNOWN
    SleepSessionRecord.STAGE_TYPE_AWAKE -> SleepStageTypeDto.AWAKE
    SleepSessionRecord.STAGE_TYPE_SLEEPING -> SleepStageTypeDto.SLEEPING
    SleepSessionRecord.STAGE_TYPE_OUT_OF_BED -> SleepStageTypeDto.OUT_OF_BED
    SleepSessionRecord.STAGE_TYPE_LIGHT -> SleepStageTypeDto.LIGHT
    SleepSessionRecord.STAGE_TYPE_DEEP -> SleepStageTypeDto.DEEP
    SleepSessionRecord.STAGE_TYPE_REM -> SleepStageTypeDto.REM
    SleepSessionRecord.STAGE_TYPE_AWAKE_IN_BED -> SleepStageTypeDto.IN_BED
    else -> SleepStageTypeDto.UNKNOWN
}

internal fun SleepStageDto.toHealthConnect(): SleepSessionRecord.Stage = SleepSessionRecord.Stage(
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    stage = stage.toHealthConnect(),
)

internal fun SleepSessionRecord.Stage.toDto(): SleepStageDto = SleepStageDto(
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    stage = stage.toSleepStageTypeDto(),
)

/**
 * Converts a Health Connect [SleepSessionRecord] object to a [SleepSessionRecordDto].
 */
internal fun SleepSessionRecord.toDto(): SleepSessionRecordDto = SleepSessionRecordDto(
    id = metadata.id,
    metadata = metadata.toDto(),
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
    endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    title = title,
    notes = notes,
    stages = stages.map { it.toDto() },
)

/**
 * Converts a [SleepSessionRecordDto] to a Health Connect [SleepSessionRecord] object.
 */
internal fun SleepSessionRecordDto.toHealthConnect(): SleepSessionRecord = SleepSessionRecord(
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    startZoneOffset = startZoneOffsetSeconds?.let {
        ZoneOffset.ofTotalSeconds(it.toInt())
    } ?: ZoneOffset.UTC,
    endZoneOffset = endZoneOffsetSeconds?.let {
        ZoneOffset.ofTotalSeconds(it.toInt())
    } ?: ZoneOffset.UTC,
    title = title,
    notes = notes,
    stages = stages.map { it.toHealthConnect() },
    metadata = metadata.toHealthConnect(),
)

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/sleep_session_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show SleepSessionRecord, HealthRecordId, sinceV1_0_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/sleep_stage_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_platform_api.g.dart'
    show SleepSessionRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [SleepSessionRecord] to [SleepSessionRecordDto].
@sinceV1_0_0
@internal
extension SleepSessionRecordToDto on SleepSessionRecord {
  SleepSessionRecordDto toDto() {
    return SleepSessionRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDto(),
      title: title,
      notes: notes,
      stages: samples.map((s) => s.toDto()).toList(),
    );
  }
}

/// Converts [SleepSessionRecordDto] to [SleepSessionRecord].
@sinceV1_0_0
@internal
extension SleepSessionRecordDtoToDomain on SleepSessionRecordDto {
  SleepSessionRecord toDomain() {
    return SleepSessionRecord(
      id: id?.toHealthRecordId() ?? HealthRecordId.none,
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDomain(),
      samples: stages.map((s) => s.toDomain()).toList(),
      title: title,
      notes: notes,
    );
  }
}
